### DISCUSSION TITLE
Getting declared base interfaces only

### DISCUSSION THREAD

**Discussion by daningalla** [icsharpcode/ILSpy#2118] [General] [answered]
Given I have a class:

`
class MyDictionary : IDictionary<string, object> {}
`

calling any of the properties or methods that discover base types such as `type.DirectBaseTypes` or `type.GetAllInterfaces()`, etc., returns not only the interfaces (interface in this case) that are declared in code, but these methods automatically flatten the hierarchy. So, in the above example I get: `IDictionary<K,V>`, `IEnumerable<KeyValuePair<K,V>>`, `ICollection<KeyValuePair<K,V>>` and  `IEnumerable`.

How do you suggest discovering just the "declared" interface implementation and not the implicit hierarchy?

Thank you.

**Comment by dgrunwald** [comment, marked as answer]
The C# compiler already flattens the hierarchy for interfaces; there is no difference in IL between
`class MyDictionary : IDictionary<string, object> {}`
and
`class MyDictionary : IDictionary<string, object>, IEnumerable<KeyValuePair<K,V>> {}`

The IL is always:
```
.class private auto ansi beforefieldinit ParallelForEachTest.MyDictionary
	extends [System.Runtime]System.Object
	implements class [System.Runtime]System.Collections.Generic.IDictionary`2<string, object>,
	           class [System.Runtime]System.Collections.Generic.ICollection`1<valuetype [System.Runtime]System.Collections.Generic.KeyValuePair`2<string, object>>,
	           class [System.Runtime]System.Collections.Generic.IEnumerable`1<valuetype [System.Runtime]System.Collections.Generic.KeyValuePair`2<string, object>>,
	           [System.Runtime]System.Collections.IEnumerable
```

`DirectBaseTypes` gives the base type and interfaces directly declared in the IL code for that class; it doesn't flatten anything more than it is already in the IL code.
